src/dev/lupio/lupio_rtc.cc

```diff
@@ -0,0 +1,128 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_rtc.hh"
+
+#include "base/time.hh"
+#include "debug/LupioRTC.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioRTC.hh"
+
+namespace gem5
+{
+
+using namespace RiscvISA;
+
+LupioRTC::LupioRTC(const Params &params) :
+    BasicPioDevice(params, params.pio_size),
+    time(params.time)
+{
+    gettimeofday(&tv, NULL);
+    gmtime_r(&tv.tv_sec, &time);
+
+    DPRINTF(LupioRTC, "Real-time clock set to %s", asctime(&time));
+}
+
+uint8_t
+LupioRTC::lupioRTCRead(uint8_t addr)
+{
+    uint8_t r = 0;
+
+    /**
+     * Get real time in seconds
+     */
+    gettimeofday(&tv, NULL);
+
+    /**
+     * Transform into broken-down time representation
+     */
+    gmtime_r(&tv.tv_sec, &time);
+
+    switch (addr) {
+        case LUPIO_RTC_SECD:
+            r = time.tm_sec;                 /* 0-60 (for leap seconds) */
+            break;
+        case LUPIO_RTC_MINT:
+            r = time.tm_min;                 /* 0-59 */
+            break;
+        case LUPIO_RTC_HOUR:
+            r = time.tm_hour;                /* 0-23 */
+            break;
+        case LUPIO_RTC_DYMO:
+            r = time.tm_mday;                /* 1-31 */
+            break;
+        case LUPIO_RTC_MNTH:
+            r = time.tm_mon + 1;             /* 1-12 */
+            break;
+        case LUPIO_RTC_YEAR:
+            r = (time.tm_year + 1900) % 100; /* 0-99 */
+            break;
+        case LUPIO_RTC_CENT:
+            r = (time.tm_year + 1900) / 100; /* 0-99 */
+            break;
+        case LUPIO_RTC_DYWK:
+            r = 1 + (time.tm_wday + 6) % 7;  /* 1-7 (Monday is 1) */
+            break;
+        case LUPIO_RTC_DYYR:
+            r = time.tm_yday + 1;            /* 1-366 (for leap years) */
+            break;
+    }
+    return r;
+}
+
+Tick
+LupioRTC::read(PacketPtr pkt)
+{
+    bool is_atomic = pkt->isAtomicOp() && pkt->cmd == MemCmd::SwapReq;
+    DPRINTF(LupioRTC,
+        "Read request - addr: %#x, size: %#x, atomic:%d\n",
+        pkt->getAddr(), pkt->getSize(), is_atomic);
+
+    Addr rtc_addr = pkt->getAddr() - pioAddr;
+    uint8_t time_val = lupioRTCRead(rtc_addr);
+    DPRINTF(LupioRTC, "time: %d\n", time_val);
+
+    pkt->setData(&time_val);
+
+    if (is_atomic) {
+        pkt->makeAtomicResponse();
+    } else {
+        pkt->makeResponse();
+    }
+
+    return pioDelay;
+}
+
+Tick
+LupioRTC::write(PacketPtr pkt)
+{
+    panic("Unexpected write to the LupioRTC device!");
```

Comment:
Are you sure that panicking is the right way to go? Isn't there a more gentle way to go about it: like print a warning and/or return some error to the CPU as part of the simulation?

Reply:
I think this is the right way to do it. There's two ways to look at it.
1. If the simulator has a bug and delivers a write to this device, we want to panic, it's a bug.
2. If a student is writing a device driver and does something wrong, rather than having to write good error handling code and checking the error correctly, I think we want to "fail fast" and have the simulator terminate. This is, in some sense, causing a "puff of smoke" out of the hardware.

Reply:
A bus error would be a property of the bus protocol, right? What is the bus protocol for the RTC? It's just wires. There's no way to signal an error without having some kind of higher-level protocol in place. There's no default protocol in gem5 (or in any hardware). 

We can certainly just ignore writes, but that could mask bugs. I still think panic-ing (as defined by gem5) is the right thing to do. However, if you have an example where this device is written and it's not either a hardware or software bug, I can change my mind :)

Reply:
Fine by me if it's the right way to go :+1: